spring-cloud-gateway-core/src/main/java/org/springframework/cloud/gateway/filter/factory/RetryGatewayFilterFactory.java

```diff
@@ -123,16 +123,19 @@ public GatewayFilter apply(RetryConfig retryConfig) {
 					return false;
 				}
 
-				for (Class<? extends Throwable> clazz : retryConfig.getExceptions()) {
-					if (clazz.isInstance(context.exception())) {
-						trace("exception is retryable %s, configured exceptions",
-								context.exception().getClass().getName(),
+				Throwable exception = context.exception();
+				for (Class<? extends Throwable> retryableClass : retryConfig
+						.getExceptions()) {
+					if (retryableClass.isInstance(exception) || (exception != null
+							&& retryableClass.isInstance(exception.getCause()))) {
+						trace("exception or its cause is retryable %s, configured exceptions %s",
+								getExceptionNameWithCause(exception),
 								retryConfig.getExceptions());
 						return true;
 					}
 				}
-				trace("exception is not retryable %s, configured exceptions",
-						context.exception().getClass().getName(),
+				trace("exception or its cause is not retryable %s, configured exceptions %s",
+						getExceptionNameWithCause(exception),
```

Comment:
Better to replace with supplier, so that exception name is constructed only when trace log level is enabled

Reply:
Did you mean `private void trace(Supplier<String> message, Supplier<Object>... args)` ?